### PR TITLE
Temoa

### PR DESCRIPTION
**Pull request by samgdotson**
This PR:
- Updates all temoa models to match the current version of temoa.
- Adds items to .gitignore
- Makes ``data_parser.py`` more robust to future changes by adding regex
- Adds packages to ``requirements.txt`` to match the temoa environment

New temoa models can now be added. 
Eventually, these first scenarios will be removed from the snakemake once we have a complete model of UIUC that includes transportation, energy storage, and new economies like hydrogen.

### CODE REVIEW

File: temoa-uiuc/Snakefile [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
-#           "02_uiuc_scenario02_model",
-#           "03_uiuc_scenario03_model",
-#           "04_uiuc_scenario04_model"]
+DBS = ["bau_uiuc", "01_uiuc", "02_uiuc", "03_uiuc", "04_uiuc"]
+SCENARIOS = ["bau", "scenario01", "scenario02", "scenario03", "scenario04"]
+FOLDER = ["bau_uiuc_bau_model",
+          "01_uiuc_scenario01_model",
+          "02_uiuc_scenario02_model",
+          "03_uiuc_scenario03_model",
+          "04_uiuc_scenario04_model"]
```
Comment (yardasol) [inline temoa-uiuc/Snakefile]: Scenario 4 is now working I assume?

File: README.md [hunk 2]
```diff
@@ -53,6 +53,8 @@ cd temoa-uiuc
 source activate temoa-py3
 pip install snakemake
 snakemake --cores=4
+# if the build fails due to file system latency, try
+# snakemake --cores=4 --latency-wait=10
```
Comment (yardasol) [inline README.md:57]: nice tip

File: .gitignore [hunk 3]
```diff
@@ -2,6 +2,8 @@
 __pycache__
 .ipynb_checkpoints
 .snakemake
+*.swp
+*.pyc
```
Comment (yardasol) [inline .gitignore:6]: These two lines are to keep the git repo free of clutter from generated files I assume?

File: requirements.txt [hunk 4]
```diff
@@ -1,6 +1,15 @@
+xlwt
```
Comment (yardasol) [inline requirements.txt:1]: Checks out with the [.yml](https://github.com/TemoaProject/temoa/blob/energysystem/environment.yml) file in the Temoa repo.

File: temoa-uiuc/data_parser.py [hunk 5]
```diff
@@ -284,7 +284,7 @@ def bar_plot(dataframe, variable, scenario, sector, emission=None, save=True):
         ax = dataframe.loc[1:, dataframe.columns != 'total'].plot.bar()
         plt.suptitle(
             (f"{scenario.upper()}: Total Annual {emission.upper()} in"
-            f" {units[variable.lower()]}"),
+             f" {units[variable.lower()]}"),
```
Comment (yardasol) [inline temoa-uiuc/data_parser.py:287]: nice spacing

File: temoa-uiuc/data_parser.py [hunk 6]
```diff
@@ -27,6 +35,11 @@ def data_by_year(datalines, year):
     This function takes in a list of datalines and returns
     a new list of datalines for a specified year.
 
+    NOTE: This function picks out the year based on a
```
Comment (yardasol) [inline temoa-uiuc/data_parser.py:38]: good note

File: temoa-uiuc/data_parser.py [hunk 7]
```diff
@@ -279,7 +295,8 @@ def bar_plot(dataframe, variable, scenario, sector, emission=None, save=True):
                            'total'].plot.bar(stacked=True)
         bars = ax.patches
         plt.suptitle(
-            f"{scenario.upper()}: Total Annual {variable} in {units[variable.lower()]}",
+            (f"{scenario.upper()}: Total Annual {variable} in"
+             f"{units[variable.lower()]}"),
```
Comment (yardasol) [inline temoa-uiuc/data_parser.py:299]: good formatting

File: temoa-uiuc/data_parser.py [hunk 8]
```diff
@@ -300,7 +317,8 @@ def bar_plot(dataframe, variable, scenario, sector, emission=None, save=True):
     if save is True:
         if emission is not None:
             plt.savefig(
-                f"{target_folder}{scenario}_{sector}_{variable.lower()}_{emission}.png")
+                (f"{target_folder}{scenario}_{sector}_{variable.lower()}_"
+                 f"{emission}.png"))
```
Comment (yardasol) [inline temoa-uiuc/data_parser.py:321]: good formatting

File: temoa-uiuc/data_parser.py [hunk 9]
```diff
@@ -158,6 +172,7 @@ def create_column(lines, years, tech):
     """
     column = {"Year": years, tech: []}
     tech_data = data_by_tech(lines, tech)
+
```
Comment (yardasol) [inline temoa-uiuc/data_parser.py:175]: nice spacing

File: temoa-uiuc/data_parser.py [hunk 10]
```diff
@@ -100,8 +114,8 @@ def data_by_tech(datalines, tech):
         data for a particular tech.
     """
     datatech = []
-
     for line in datalines:
+        # print(line)
```
Comment (yardasol) [inline temoa-uiuc/data_parser.py:118]: can probably remove this

File: temoa-uiuc/data_parser.py [hunk 11]
```diff
@@ -40,11 +53,12 @@ def data_by_year(datalines, year):
         This is a list of datalines that only contains
         data for a particular year.
     """
-    string = f"[{year},"
     datayear = []
 
     for line in datalines:
-        if string in line:
+        line_year = re.findall(r"[-+]?\d*\.\d+|\d+", line)[1]
```
Comment (yardasol) [inline temoa-uiuc/data_parser.py:59]: It seems like you are searching for regular expressions that match the pattern of a number with arbitrary number of digits before the decimal and after the decimal, OR just a number with no decimals and an arbitrary number of digits.

You then check if this regular expression when converted to an integer equals our 'year' parameter.

I think this should work fine.

File: temoa-uiuc/data_files/01_uiuc.sql [hunk 12]
```diff
@@ -1,48 +1,20 @@
 BEGIN TRANSACTION;
-
```
Comment (yardasol) [inline temoa-uiuc/data_files/01_uiuc.sql:2]: I found the comments in this iteration of the .sql file to be very helpful when reading through it.

File: temoa-uiuc/data_files/01_uiuc.sql [hunk 13]
```diff
@@ -1,48 +1,20 @@
...
-INSERT INTO "sector_labels" VALUES('commercial');
-INSERT INTO "sector_labels" VALUES('university');
-INSERT INTO "sector_labels" VALUES('industrial');
-
---Tables below correspond to Temoa sets
-CREATE TABLE time_periods (
-  t_periods integer primary key,
-  flag text,
-  FOREIGN KEY(flag) REFERENCES time_period_labels(t_period_labels));
+CREATE TABLE "time_season" (
```
Comment (yardasol) [inline temoa-uiuc/data_files/01_uiuc.sql:2]: Perhaps consider adding similar comments in the current iteration?

File: temoa-uiuc/data_files/02_uiuc.sql [hunk 14]
```diff
@@ -1,48 +1,20 @@
 BEGIN TRANSACTION;
-
```
Comment (yardasol) [inline temoa-uiuc/data_files/02_uiuc.sql:2]: I found the comments in this iteration of the .sql file to be very helpful when reading through it.

File: temoa-uiuc/data_files/02_uiuc.sql [hunk 15]
```diff
@@ -1,48 +1,20 @@
...
-INSERT INTO "sector_labels" VALUES('commercial');
-INSERT INTO "sector_labels" VALUES('university');
-INSERT INTO "sector_labels" VALUES('industrial');
-
---Tables below correspond to Temoa sets
-CREATE TABLE time_periods (
-  t_periods integer primary key,
-  flag text,
-  FOREIGN KEY(flag) REFERENCES time_period_labels(t_period_labels));
+CREATE TABLE "time_season" (
```
Comment (yardasol) [inline temoa-uiuc/data_files/02_uiuc.sql:2]: Perhaps consider adding similar comments in current iteration?

File: temoa-uiuc/data_files/03_uiuc.sql [hunk 16]
```diff
@@ -1,48 +1,20 @@
 BEGIN TRANSACTION;
-
```
Comment (yardasol) [inline temoa-uiuc/data_files/03_uiuc.sql:2]: I found the comments in this iteration of the .sql file to be very helpful when reading through it.

File: temoa-uiuc/data_files/03_uiuc.sql [hunk 17]
```diff
@@ -1,48 +1,20 @@
...
-INSERT INTO "sector_labels" VALUES('commercial');
-INSERT INTO "sector_labels" VALUES('university');
-INSERT INTO "sector_labels" VALUES('industrial');
-
---Tables below correspond to Temoa sets
-CREATE TABLE time_periods (
-  t_periods integer primary key,
-  flag text,
-  FOREIGN KEY(flag) REFERENCES time_period_labels(t_period_labels));
+CREATE TABLE "time_season" (
```
Comment (yardasol) [inline temoa-uiuc/data_files/03_uiuc.sql:2]: Perhaps consider adding similar comments in the current iteration?

File: temoa-uiuc/data_files/04_uiuc.sql [hunk 18]
```diff
@@ -1,48 +1,20 @@
 BEGIN TRANSACTION;
-
```
Comment (yardasol) [inline temoa-uiuc/data_files/04_uiuc.sql:2]: I found the comments in this iteration of the .sql file to be very helpful when reading through it.

File: temoa-uiuc/data_files/04_uiuc.sql [hunk 19]
```diff
@@ -1,48 +1,20 @@
...
-INSERT INTO "sector_labels" VALUES('commercial');
-INSERT INTO "sector_labels" VALUES('university');
-INSERT INTO "sector_labels" VALUES('industrial');
-
---Tables below correspond to Temoa sets
-CREATE TABLE time_periods (
-  t_periods integer primary key,
-  flag text,
-  FOREIGN KEY(flag) REFERENCES time_period_labels(t_period_labels));
+CREATE TABLE "time_season" (
```
Comment (yardasol) [inline temoa-uiuc/data_files/04_uiuc.sql:2]: Perhaps consider adding similar comments in the current iteration?

File: temoa-uiuc/data_files/bau_uiuc.sql [hunk 20]
```diff
@@ -1,48 +1,20 @@
 BEGIN TRANSACTION;
-
```
Comment (yardasol) [inline temoa-uiuc/data_files/bau_uiuc.sql:2]: I found the comments in this iteration of the .sql file to be very helpful when reading through it.

File: temoa-uiuc/data_files/bau_uiuc.sql [hunk 21]
```diff
@@ -1,48 +1,20 @@
...
-INSERT INTO "sector_labels" VALUES('commercial');
-INSERT INTO "sector_labels" VALUES('university');
-INSERT INTO "sector_labels" VALUES('industrial');
-
---Tables below correspond to Temoa sets
-CREATE TABLE time_periods (
-  t_periods integer primary key,
-  flag text,
-  FOREIGN KEY(flag) REFERENCES time_period_labels(t_period_labels));
+CREATE TABLE "time_season" (
```
Comment (yardasol) [inline temoa-uiuc/data_files/bau_uiuc.sql:2]: Perhaps consider adding similar comments in the current iteration?

File: temoa-uiuc/Snakefile [hunk 22]
```diff
@@ -21,9 +21,9 @@ rule all:
         expand(r"data_files/{db}_{scenario}_model/{scenario}_OutputLog.log",
                zip,
                db=DBS, scenario=SCENARIOS),
-        # expand(r"data_files/{db}_{scenario}_model/{scenario}.lp",
-        #        zip,
-        #        db=DBS, scenario=SCENARIOS),
+        expand(r"data_files/{db}_{scenario}_model/{scenario}.lp",
```
Comment (yardasol) [inline temoa-uiuc/Snakefile:24]: Tell snakemake to take in .lp file files?

**Review by yardasol — COMMENTED** [review]
See individual comments in the code.

File: temoa-uiuc/data_parser.py (hunk 11 above)
Comment (samgdotson) [reply temoa-uiuc/data_parser.py:59]: If the line of data is:
```
30.086727718839995    V_EmissionActivityByPeriodAndProcess[uiuc,2026,ewaste,IMPSOL,2022]
```
This regex line will return 
```
['30.086727718839995', '2026', '2022']
```
Where the first "year" value is the year we're interested in (thus index [1]) and the second "year" value is the vintage of the technology in Temoa.

File: temoa-uiuc/data_parser.py (hunk 9 above)
Comment (samgdotson) [reply temoa-uiuc/data_parser.py:175]: This is due to the pep8 style guide. Running
```
autopep8 --in-place --aggressive --aggressive <file.py>
```
will automatically fix pep8 issues for you.

File: temoa-uiuc/Snakefile (hunk 22 above)
Comment (samgdotson) [reply temoa-uiuc/Snakefile:24]: Temoa builds .lp files. Snakemake executes a temoa run and knows its finished by the presence of a .lp file in the correct directory.

File: .gitignore (hunk 3 above)
Comment (samgdotson) [reply .gitignore:6]: That is correct

File: temoa-uiuc/Snakefile (hunk 1 above)
Comment (samgdotson) [reply temoa-uiuc/Snakefile]: Yes, it's no longer experimental

File: temoa-uiuc/data_files/04_uiuc.sql (hunk 19 above)
Comment (samgdotson) [reply temoa-uiuc/data_files/04_uiuc.sql:2]: This would be a good issue to create!

**Comment by robfairh** [comment]
I've run one of the cases and I got this warning:

Warning: ('uiuc', 'IMPELC', 2000) has a specified Efficiency, but does not have any existing install base (ExistingCapacity).
Warning: ('uiuc', 'ABBOTT', 2020) has a specified Efficiency, but does not have any existing install base (ExistingCapacity).
Warning: ('uiuc', 'TURBINE', 2020) has a specified Efficiency, but does not have any existing install base (ExistingCapacity).
Warning: ('uiuc', 'NUCLEAR', 2020) has a specified Efficiency, but does not have any existing install base (ExistingCapacity).
Warning: ('uiuc', 'UL', 2000) has a specified Efficiency, but does not have any existing install base (ExistingCapacity).
Warning: ('uiuc', 'UL', 2020) has a specified Efficiency, but does not have any existing install base (ExistingCapacity).
Warning: ('uiuc', 'UH', 2000) has a specified Efficiency, but does not have any existing install base (ExistingCapacity).
Warning: ('uiuc', 'UH', 2020) has a specified Efficiency, but does not have any existing install base (ExistingCapacity).

I am still familiarizing myself with TEMOA, how important is this warning?

**Comment by samgdotson** [comment]
I don't think it's a problem, it still produces results. I'll check if it changes anything.

**Comment by samgdotson** [comment]
@robfairh
1) The warning doesn't hurt anything. It's just pointing out that I offered an efficiency for some technologies that didn't actually exist in the years specified. 2000 and 2020 are considered "existing" time periods. We didn't build an import system (IMPELC) in 2000 so the technology doesn't "exist". It's a bit weird. 
2) I have fixed scenario04 to stop the warnings. If we want to do this for the other scenarios we can (by opening a new issue) but it's not essential.

**Comment by robfairh** [comment]
Great! I think you have addressed all of @yardasol's comments as well. So I'll merge the PR.

**Review by robfairh — APPROVED** [review]

**Pull request merged**